### PR TITLE
Enhance/student import named columns fixed

### PR DESCRIPTION
**Pull request by lachlanwest**
Fixed version of previous branch. Should not be any old code tagging along.

### CODE REVIEW

File: test_files/COS20007-Enrol-Students.csv [hunk 1]
```diff
@@ -1 +1,13 @@
...
+COS20007,student_2,Emmanuel,Ondricka,student_2@doubtfire.com,LA1-01
+COS20007,student_3,Hettie,Hermiston,student_3@doubtfire.com,LA1-03
+COS20007,student_4,Dolly,Leuschke,student_4@doubtfire.com,LA1-03
+COS20007,student_5,Dedric,Brekke,student_5@doubtfire.com,LA1-03
+COS20007,cliff,Cliff,Warren,cliff@doubtfire.com,LA1-03
+COS20007,student_99,New Student,Name,student_99@swin.edu.au,LA1-01
+COS20007,student_100,Student 100,S100 Name,student_100@swin.edu.au,LA1-99
+COS20007,student_101,Student 101,S101 Name,student_101@swin.edu.au,
+COS20007,student_102,Email,Wrong,student_99@swin.edu.au,LA1-01
+COS20007,cliff,Cliff,Warren,cliff@doubtfire.com,LA1-03
```
Comment (macite) [inline test_files/COS20007-Enrol-Students.csv]: Keep the empty rows in here to test how that works when empty rows are imported.

File: app/models/unit.rb [hunk 2]
```diff
@@ -223,23 +223,25 @@ def tutorial_with_abbr(abbr)
 
   #
   # Imports users into a project from CSV file.
-  # Format: unit_code, Student ID,First Name,Surname,email,tutorial
-  #
+  # Format: Subject/Unit Code, Student ID,First Name, Surname, email, tutorial
```
Comment (macite) [inline app/models/unit.rb]: As this now needs it to be unit_code you could update the format here.

File: app/models/unit.rb [hunk 3]
```diff
@@ -223,23 +223,25 @@ def tutorial_with_abbr(abbr)
...
+  # Expected columns: unit_code, username, first_name, last_name, email, tutorial
   def import_users_from_csv(file)
     tutorial_cache = {}
     success = []
     errors = []
     ignored = []
 
-    CSV.foreach(file) do |row|
+    CSV.foreach(file, headers: true) do |row|
```
Comment (macite) [inline app/models/unit.rb]: Also add in the following to ensure that empty and strange UTF8 characters are handled ok...

```:header_converters => [lambda { |i| i.nil? ? '' : i },  :downcase, lambda { |hdr| hdr.strip unless hdr.nil?}],
        :converters => [lambda{ |body| body.encode!('UTF-8', 'binary', invalid: :replace, undef: :replace, replace: '') unless body.nil? }]```

File: app/models/unit.rb [hunk 4]
```diff
@@ -223,23 +223,25 @@ def tutorial_with_abbr(abbr)
...
       # Make sure we're not looking at the header or an empty line
-      next if row[0] =~ /(subject|unit)_code/
-      # next if row[5] !~ /^LA\d/
+      next if row['unit_code'] =~ /unit_code/
```
Comment (macite) [inline app/models/unit.rb]: Add a check here for missing columns:

```missing = missing_headers(row, ['group_name', 'username', 'tutorial'])
        if missing.count > 0
          errors << { row: row, message: "Missing headers: #{missing.join(', ')}" }
          next
        end```

**Comment by alexcu** [comment]
@lachlanwest What was the status of this branch?

**Comment by macite** [comment]
@alexcu This is still open, waiting for the above comments to be addressed. Not sure what progress @lachlanwest has made since then.

**Comment by alexcu** [comment]
Didn't realise there was still a PR open for this. I will restore the branch

**Comment by alexcu** [comment]
@lachlanwest will be focusing on the web front-end for the ticketing system so perhaps we close this for now @macite and review at a later stage?

**Pull request merged**